### PR TITLE
Migration to nns.internetcomputer.org

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

We need to move from `nns.ic0.app` to `nns.internetcomputer.org`.

The requests should move from `ic0.app` to `ic-api.io`.

# Changes

* Add `.well-known` assets to use custom domain `nns.internetcomputer.org`.
* Rename `nns.ic0.app` to `nns.internetcomputer.org`.
* Rename host from `ic0.app` to `ic-api.io`.
* Change identity provider in mainnet to `identity.internetcomputer.org`.
* Add a rule to use `identity.ic0.app` in case the user lands in `nns.ic0.app`.
* Add new urls to CSP rules.

# Tests

* Test the specific rule that should return the old identity provider url.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@mstrasinskis @peterpeterparker @bitdivine please review

I deployed it to the nnsdapp: https://xub3y-eqaaa-aaaaa-aaawq-cai.nnsdapp.testnet.dfinity.network/

**Comment by lmuntaner** [comment]
@frederikrothenberger maybe it would be good if you can check that we are not missing anything for the custom domain

🙏

File: dfx.json [hunk 1]
```diff
@@ -223,7 +223,7 @@
           "ENABLE_SNS_2": false,
           "ENABLE_SNS_VOTING": false,
           "ENABLE_SNS_AGGREGATOR": false,
-          "ENABLE_CKBTC_LEDGER": false
```
Comment (frederikrothenberger) [inline dfx.json:226]: Is this related to the migration?

File: dfx.json [hunk 2]
```diff
@@ -223,7 +223,7 @@
...
+          "ENABLE_CKBTC_LEDGER": true
```
Comment (mstrasinskis) [inline dfx.json:226]: Should it be enabled?

File: frontend/scripts/build.csp.mjs [hunk 3]
```diff
@@ -144,6 +144,8 @@ const cspConnectSrc = () => {
   // TODO: Use `URL` to check if the URL is valid and not introduce a security issue
   const src = [
     process.env.VITE_IDENTITY_SERVICE_URL,
+    // We move to internetcomputer.org, but if a user access the app with the old URL, we need to allow it
+    "https://identity.ic0.app",
```
Comment (frederikrothenberger) [inline frontend/scripts/build.csp.mjs:148]: Shouldn't this be `https://nns.ic0.app`?

File: frontend/src/lib/stores/auth.store.ts [hunk 4]
```diff
@@ -55,7 +75,7 @@ const initAuthStore = () => {
       authClient = authClient ?? (await createAuthClient());
 
       await authClient?.login({
-        identityProvider: IDENTITY_SERVICE_URL,
+        identityProvider: getIdentityProvider(),
```
Comment (frederikrothenberger) [inline frontend/src/lib/stores/auth.store.ts:78]: You need to start on passing in the `derivationOrigin: https://nns.ic0.app`.

File: frontend/src/lib/constants/environment.constants.ts [hunk 5]
```diff
@@ -32,7 +32,7 @@ export const {
 export const IS_TESTNET: boolean =
   DFX_NETWORK !== "mainnet" &&
   FETCH_ROOT_KEY === true &&
-  !HOST.includes(".ic0.app");
+  !HOST.includes(".ic-api.io");
```
Comment (mstrasinskis) [inline frontend/src/lib/constants/environment.constants.ts:35]: Shouldn't both be supported for some time? Asking because of this comment https://github.com/dfinity/nns-dapp/pull/1895/files#diff-ef8663875b35cfc44789b6c8e8004e0064a2fb43f337f792126deeef55529c92R147

**Review by frederikrothenberger — COMMENTED** [review]
I'm missing a lot of context. Left some comments. Definitely have somebody more familiar with the NNS dapp review it as well.

File: frontend/src/lib/rest/governance-metrics.rest.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 // TODO: Extract .env
 const GOVERNANCE_CANISTER_RAW_URL =
-  "https://rrkah-fqaaa-aaaaa-aaaaq-cai.raw.ic0.app";
+  "https://rrkah-fqaaa-aaaaa-aaaaq-cai.raw.ic-api.io";
```
Comment (nmattia) [inline frontend/src/lib/rest/governance-metrics.rest.ts:3]: I think you want `icp0.io` here, `ic-api.io` (AFAICT) would only be for canister calls

File: frontend/src/lib/stores/auth.store.ts [hunk 7]
```diff
@@ -16,6 +17,25 @@ export interface AuthStore {
 // @see agent-js issue [#618](https://github.com/dfinity/agent-js/pull/618)
 let authClient: AuthClient | undefined | null;
 
+const getIdentityProvider = () => {
+  const environmentUrl = IDENTITY_SERVICE_URL;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const location = (window as any)?.location;
```
Comment (peterpeterparker) [inline frontend/src/lib/stores/auth.store.ts:24]: - `if (!browser) { return .... }` maybe?

- is the cast to `any` a must?

File: rs/backend/src/assets.rs [hunk 8]
```diff
@@ -218,6 +218,10 @@ pub fn http_request(req: HttpRequest) -> HttpResponse {
                     if let Some(content_encoding_header) = content_encoding.header() {
                         headers.push(("Content-Encoding".to_string(), content_encoding_header.to_string()));
                     }
+                    // Assets within .well-known are used by II and should be accessible
+                    if request_path.starts_with("/.well-known") {
+                        headers.push(("Access-Control-Allow-Origin".to_string(), "*".to_string()));
```
Comment (peterpeterparker) [inline rs/backend/src/assets.rs:223]: @frederikrothenberger `*` is fine or should be more restrictive in case of nns-dapp?

Comment (frederikrothenberger) [reply rs/backend/src/assets.rs:223]: It is fine if only added to the `/.well-knonw` path.

Otherwise this needs to be either
* https://identity.ic0.app
* https://identity.internetcomputer.org
depending on which URL you are pointing to for authentication.

File: dfx.json (hunk 1 above)
Comment (lmuntaner) [reply dfx.json:226]: Not, but needed to test the release candidate with prod-like functionality. It's enabled for `mainet`.

This is fixed in `main`, but I started this branch from `prod` tag.

File: dfx.json (hunk 2 above)
Comment (lmuntaner) [reply dfx.json:226]: Yes, `staging` should be a copy of `mainnet`.

This is fixed in main, but I started this branch from prod tag.

File: frontend/scripts/build.csp.mjs (hunk 3 above)
Comment (lmuntaner) [reply frontend/scripts/build.csp.mjs:148]: No, but I should also add it.

The OWN_CANISTER_URL is that one.

File: frontend/src/lib/constants/environment.constants.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/lib/constants/environment.constants.ts:35]: yes, probably, good point

File: frontend/src/lib/rest/governance-metrics.rest.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/rest/governance-metrics.rest.ts:3]: Isn't

```
await fetch(`${GOVERNANCE_CANISTER_RAW_URL}/metrics`)
```

A canister call?

File: frontend/src/lib/stores/auth.store.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/stores/auth.store.ts:78]: Do I need this for testnet? or only for mainnet?

File: dfx.json (hunk 2 above)
Comment (mstrasinskis) [reply dfx.json:226]: Thanks for the explanation!

File: frontend/src/lib/stores/auth.store.ts (hunk 4 above)
Comment (nmattia) [reply frontend/src/lib/stores/auth.store.ts:80]: Only for mainnet, I believe

File: frontend/src/lib/rest/governance-metrics.rest.ts (hunk 6 above)
Comment (nmattia) [reply frontend/src/lib/rest/governance-metrics.rest.ts:3]: no, that looks like a good old http fetch! although it's then turned into a canister call by the boundary nodes, but you shouldn't have to worry about that

File: frontend/jest-setup.ts [hunk 9]
```diff
@@ -34,7 +34,7 @@ jest.mock("./src/lib/constants/canister-ids.constants.ts", () => ({
 
 jest.mock("./src/lib/constants/environment.constants.ts", () => ({
   DFX_NETWORK: "testnet",
-  HOST: "https://ic0.app",
+  HOST: "https://nns.internetcomputer.org",
```
Comment (nmattia) [inline frontend/jest-setup.ts:37]: Might be wrong (no experience with the codebase) but I think you want `ic-api.io`; `HOST` seems to be the host used for api calls, not the host the assets are served from

File: frontend/scripts/build.csp.mjs (hunk 3 above)
Comment (nmattia) [reply frontend/scripts/build.csp.mjs:148]: I'm always a bit fuzzy on CSP issues, but do you really need II's URL in there? do you issue calls to II?

File: frontend/src/lib/constants/environment.constants.ts [hunk 10]
```diff
@@ -32,7 +34,7 @@ export const {
 export const IS_TESTNET: boolean =
   DFX_NETWORK !== "mainnet" &&
   FETCH_ROOT_KEY === true &&
-  !HOST.includes(".ic0.app");
+  !(HOST.includes(".ic-api.io") || HOST.includes(".ic0.app"));
```
Comment (nmattia) [inline frontend/src/lib/constants/environment.constants.ts:37]:
```suggestion
  !(HOST.includes(".icp-api.io") || HOST.includes(".ic0.app"));
```

File: frontend/src/lib/stores/auth.store.ts [hunk 11]
```diff
@@ -16,6 +22,22 @@ export interface AuthStore {
 // @see agent-js issue [#618](https://github.com/dfinity/agent-js/pull/618)
 let authClient: AuthClient | undefined | null;
 
+const getIdentityProvider = () => {
+  const environmentUrl = IDENTITY_SERVICE_URL;
```
Comment (nmattia) [inline frontend/src/lib/stores/auth.store.ts:26]: nit: why not use `IDENTITY_SERVICE_URL` directly?

File: frontend/scripts/build.csp.mjs (hunk 3 above)
Comment (lmuntaner) [reply frontend/scripts/build.csp.mjs:148]: I honestly haven't checked. That might be worth it's own review and maybe removal.

But in this PR I didn't want to add any other change. It was already part of the CSP; therefore, I kept it.

File: frontend/src/lib/stores/auth.store.ts (hunk 4 above)
Comment (lmuntaner) [reply frontend/src/lib/stores/auth.store.ts:80]: Ok, then no. Thanks for the feedback. That's what I thought so as well.

Comment (lmuntaner) [reply frontend/src/lib/stores/auth.store.ts:80]: done, btw

File: frontend/src/lib/rest/governance-metrics.rest.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/rest/governance-metrics.rest.ts:3]: This is a little bit confusing for me. So the old canisters will still live in `icp0.io`?

We also have the call of the SNS Aggregator, which is also an HTTP request. It uses the HOST `<id>.icp0.io`.

If SNS Aggregator doesn't work is worse than these metrics or the favicons.

I'll ask around.

File: frontend/jest-setup.ts (hunk 9 above)
Comment (lmuntaner) [reply frontend/jest-setup.ts:37]: yes, true. Although this is just for testing. But I'll change it.

File: frontend/src/lib/constants/environment.constants.ts (hunk 10 above)
Comment (lmuntaner) [reply frontend/src/lib/constants/environment.constants.ts:37]: thanks!

File: frontend/src/lib/stores/auth.store.ts (hunk 11 above)
Comment (lmuntaner) [reply frontend/src/lib/stores/auth.store.ts:26]: I started with your code and refactored. This doesn't make sense anymore, I agree.

File: frontend/src/lib/rest/governance-metrics.rest.ts (hunk 6 above)
Comment (lmuntaner) [reply frontend/src/lib/rest/governance-metrics.rest.ts]: The SNS Aggregator is not yet used. Therefore, we'll ignore it for now.

I will revert these changes and use the old `ic0.app` for this call as suggested by Björn.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx
